File: Leetcode/Python/2958.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        mapping = defaultdict(int)
        left = 0
        max_len = 0
        for right in range(len(nums)):
            num = nums[right]
            mapping[num] += 1
            while mapping[num] > k:
                remove_num = nums[left]
                mapping[remove_num] -= 1
                left += 1
            max_len = max(max_len, right - left + 1)
        return max_len

File: Leetcode/cpp/2958.cpp
class Solution {
public:
    int maxSubarrayLength(std::vector<int>& nums, int k) {
        std::unordered_map<int, int> map;
        int left = 0;
        int max_len = 0;
        for (int right = 0; right < nums.size(); ++right) {
            int num = nums[right];
            ++map[num];
            while (map[num] > k) {
                int remove = nums[left++];
                --map[remove];
            }
            max_len = std::max<int>(max_len, right - left + 1);
        }
        return max_len;
    }
};
